core/src/main/scala/kafka/log/LogSegment.scala

```diff
@@ -200,15 +226,40 @@ class LogSegment(val log: FileMessageSet,
     val truncated = log.sizeInBytes - validBytes
     log.truncateTo(validBytes)
     index.trimToValidSize()
+    // A normally closed segment always appends the biggest timestamp ever seen into log segment, we do this as well.
+    if (maxTimestampSoFar >= 0)
+      timeIndex.maybeAppend(maxTimestampSoFar, offsetOfMaxTimestamp)
+    timeIndex.trimToValidSize()
     truncated
   }
 
+  def loadLargestTimestamp(readToLogEnd: Boolean = false) {
+    // Get the last time index entry. If the time index is empty, it will return (-1, baseOffset)
+    val lastTimeIndexEntry = timeIndex.lastEntry
+    if (readToLogEnd) {
+      val offsetPosition = index.lookup(lastTimeIndexEntry.offset)
+      // Scan the rest of the messages to see if there is a larger timestamp after the last time index entry.
+      maxTimestampSoFar = math.max(lastTimeIndexEntry.timestamp, log.largestTimestampAfter(offsetPosition.position))
+    } else
+      maxTimestampSoFar = lastTimeIndexEntry.timestamp
+  }
+
+  /**
+   * This method appends an entry to the time index if the time index entry is empty.
+   * The timestamp of the entry is the largest timestamp in the segment and the offset is the last offset in the segment.
+   */
+  def ensureNonEmptyTimeIndex() {
+    if (timeIndex.entries == 0)
+      timeIndex.maybeAppend(largestTimestamp, nextOffset - 1)
```

Comment:
Thought about this a little more, it seems that it would be better to append the nextOffset - 1. The offset is essentially used in two cases:
1) Search by timestamp
2) Log truncation.

For case 1, The previous behavior was that if the target timestamp is greater than the lastModified, we will simply start from the next segment. This behavior will be changed if we append the baseOffset here. i.e. user needs to start consume from the beginning of this segment instead of the beginning of next segment.

For case 2, if an inactive log is truncated and becomes the active log segment again, we will want to truncate the time index entry as well, if we append the nextOffset -1 here, the time index will also be truncated. But if we append baseOffset here, the time index entry won't be truncated.